### ISSUE TITLE
ESM contains server code, doesn't provide named exports

### ISSUE THREAD

**Issue by thekevinbrown** [diegomura/react-pdf#1317]
**Describe the bug**
When using [Vite](https://vitejs.dev/) to package a React app, we hit some issues because:

1. The ESM that `@react-pdf/renderer` provides doesn't seem to provide the same exports as the CJS module.
2. There is server code included in the ESM, but this library is a browser module.

**To Reproduce**
You can reproduce with this repository: https://github.com/thekevinbrown/vite-react-pdf-reproduction

**Expected behavior**
I expected React PDF to be able to be bundled and used in an ESM context.

**Screenshots**
```console
Uncaught ReferenceError: global is not defined
    at node_modules/blob/index.js (@react-pdf_renderer.js:554)
    at __require (chunk-KVFJW2XH.js:12)
    at node_modules/blob-stream/index.js (@react-pdf_renderer.js:616)
    at __require (chunk-KVFJW2XH.js:12)
    at @react-pdf_renderer.js:126368
```

**Desktop (please complete the following information):**
 - OS: MacOS
 - Browser: Chrome
 - React-pdf version: "@react-pdf/renderer": "^2.0.12",

**Additional Notes**

An issue was logged with Vite here: https://github.com/vitejs/vite/issues/3405

This may be related to this issue as well: #1029

I'd be happy to help with a PR if one is welcome, just wanted to check to ensure I wasn't duplicating effort and that this was something you'd accept a PR for.

**Comment by thekevinbrown** [comment]
I'm trying to even get a local development version running in an example app and I'm not sure where to start. The `examples` folder has no start script.

Does anyone know how to test changes with this library? There's no `CONTRIBUTING.md` file either, so I'm not really sure where to start.

**Comment by jeetiss** [comment]
@thekevinbrown there is [CONTRIBUTING.md](https://github.com/diegomura/react-pdf/blob/master/.github/CONTRIBUTING.md)

**Comment by thekevinbrown** [comment]
Ah, I didn't look in the `.github` folder.

Anyway, it doesn't say how to build and run the code locally, so I'm still pretty stuck. I've added Rollup to 4 of the packages in the monorepo and still can't get it to fully run in a Vite environment.

I'm thinking I'll give up and just shim `@react-pdf/renderer` at this point.

**Comment by diegomura** [comment]
Hey @thekevinbrown ! Sorry for the delay and lack of documentation on how to contribute to this project. It's really hard to catch up with all the activity here and also work haha. I'm writing a quick guide right now on the CONTRIBUTING.md file to how to run a dev environment for this project

**Comment by thekevinbrown** [comment]
@diegomura, no worries at all, I maintain some open source stuff and the struggle is real! No stress at all. The main thing that's missing IMO is a way to run the examples against the packages already in the repo so we can test changes and be sure they work. When I use a Vite environment to try to test, it breaks currently.

To give you a picture of what I'm up to, here's my WIP commit: https://github.com/TechInSite/react-pdf/tree/fix-node-code Basically I'm trying to achieve two things:

1. Provide ESM for all packages so more modern build environments can use them all easily. The fact that some are ESM and some are CJS confuses ESBuild.
2. Remove the reference to `blob-stream` as it references `blob` which has node code that references `global` in it. I suspect we don't even need it as it's used in one place to load the whole stream into an in memory Blob anyway, so I'm thinking we can just read the stream and pipe to the blob ourselves there.

The error I'm currently getting from Rollup on `packages/textkit` is: 
```
Error: 'default' is not exported by node_modules/@babel/runtime/helpers/defineProperty.js, imported by src/engines/linebreaker/linkedList.js
```

That's because I've added Rollup to that package to try to achieve point 1 above. I'm not sure if we still need Babel Runtime in that package, but if we do, not sure how to get it to work with Rollup.

I'll keep poking at it, but I'll also publish my shim so I can unblock the team while we investigate.

**Comment by thekevinbrown** [comment]
Alrighty, finally found a set of shims and workarounds that make it work, both in Vite Dev and Production builds.

- Aliases need to be created to some browserify shims for node globals.
- Because the load order is different in dev and production, I needed to put quite a few shims in different files. (Actually on disk in `node_modules`. Yuck.)
- `readable-stream` which is needed for Node streams in the browser works fine in dev mode, but in a production build it fails because of circular imports which Rollup doesn't handle particularly well, but also probably shouldn't be there in the first place in ESM land (https://github.com/rollup/rollup-plugin-commonjs/issues/293, https://github.com/nodejs/readable-stream/issues/348). I changed the package to use a Registry pattern so it'd work (https://github.com/techinsite/readable-stream) and published it as [vite-compatible-readable-stream](https://www.npmjs.com/package/vite-compatible-readable-stream).
- Because there are about 7 new dependencies needed, including Vite aliases, and manual patching of files in `node_modules` I bundled that up into [vite-plugin-shim-react](https://www.npmjs.com/package/vite-plugin-shim-react-pdf).

So, if you're coming here from Google, you do the following in your Vite project:

```console
$ yarn add vite-plugin-shim-react
```

Then in `vite.config.js`:
```javascript
import { defineConfig } from "vite";
import reactRefresh from "@vitejs/plugin-react-refresh";
import shimReactPdf from "vite-plugin-shim-react-pdf";

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [reactRefresh(), shimReactPdf()],
});
```
The you'll also need to use the default import because of the missing exports, e.g. like the test harness below does. E.g.

Instead of:
```javascript
import { Text, View, StyleSheet } from "@react-pdf/renderer";
```
You have to do:
```javascript
import pdf from "@react-pdf/renderer";
const { Text, View, StyleSheet } = pdf;
```

React PDF also didn't have a development project you could use to test changes, so I made a really rough one out of the examples: https://github.com/TechInSite/react-pdf-development-harness

@diegomura I'd be happy to transfer ownership of the development harness repo to you if you want to have it as part of the React PDF project moving forward. Could use quite a bit of CSS work, but I had enough to get through with the above that I didn't put any time into making it pretty at all.

**Comment by endlesstravel** [comment]
same issue

**Comment by carlobeltrame** [comment]
I'm having the same issue as well, and the shim of @thekevinbrown worked wonders, even for my special case of packing the react component into a webcomponent and using it inside a Vue.js app! Thanks a lot for that! Let's hope this can be fixed directly in react-pdf in the future.

**Comment by flahtorres** [comment]
I'm have the same issue, I was generating a pdf using pdfRenderer in React and the @thekevinbrown solution works, I'm very thankful =D

Anyway, the react-pdf needs to fix the problem

**Comment by carlobeltrame** [comment]
I worked around the import issue by adding a file "reactPdf.js" like follows to my project that "converts" the default export to named exports:
```js
import pdf from '@react-pdf/renderer'

export const StyleSheet = pdf.StyleSheet
export const Font = pdf.Font

// https://react-pdf.org/components
export const Document = pdf.Document
export const Page = pdf.Page
export const Image = pdf.Image
export const View = pdf.View
export const Text = pdf.Text
export const Link = pdf.Link
export const Note = pdf.Note
export const Canvas = pdf.Canvas

// https://react-pdf.org/svg
export const Svg = pdf.Svg
export const Line = pdf.Line
export const Polyline = pdf.Polyline
export const Polygon = pdf.Polygon
export const Path = pdf.Path
export const Rect = pdf.Rect
export const Circle = pdf.Circle
export const Ellipse = pdf.Ellipse
// export const Text = pdf.Text
export const Tspan = pdf.Tspan
export const G = pdf.G
export const Stop = pdf.Stop
export const Defs = pdf.Defs
export const ClipPath = pdf.ClipPath
export const LinearGradient = pdf.LinearGradient
export const RadialGradient = pdf.RadialGradient
```

And then I can import from that file normally:
```js
import { Document, Page, Text } from './reactPdf.js'
```

This always loads all the primitives from react-pdf, but as @diegomura stated in #1119, that is not a big concern, since they are mostly just plain strings.

**Comment by pmw1407** [comment]
> 

It doesn't work now ㅠㅠ

**Comment by thekevinbrown** [comment]
Can you provide an example repo so we can have a look?

**Comment by lostintheway** [comment]
> ```js
> import pdf from "@react-pdf/renderer";
> const { Text, View, StyleSheet } = pdf;
> ```

https://github.com/lostintheway/reactpdf-error

**Comment by lostintheway** [comment]
> Can you provide an example repo so we can have a look?

Here it is:
https://github.com/lostintheway/reactpdf-error

**Comment by lostintheway** [comment]
I realized I had a bug on import implementation, fixed it, thankss

**Comment by diegomura** [comment]
@thekevinbrown can we close this?

**Comment by carlobeltrame** [comment]
I don't think so. As far as I am aware, react-pdf is still broken in web bundlers newer than Webpack 4, which hasn't seen any development in almost a year now: https://github.com/webpack/webpack/tree/webpack-4

The fixes from the shim in [index.js](https://github.com/exogee-technology/vite-plugin-shim-react-pdf/blob/main/src/index.js) and [shim-react-pdf.js](https://github.com/exogee-technology/vite-plugin-shim-react-pdf/blob/main/scripts/shim-react-pdf.js) should be applied directly in the react-pdf code base, while making sure that server side operation of react-pdf still works. I can open a PR if @thekevinbrown does not have time for it, but I do not have the know-how to test whether it still works in a node.js environment.

**Comment by thekevinbrown** [comment]
Yeah, and even with my shim it still doesn't provide named exports and you still have to do:

```typescript
import pdf from "@react-pdf/renderer";
const { Text, View, StyleSheet } = pdf;
```

I'd say we should break this into two pieces that are separate issues:

1. Fixing the named exports (should be relatively easy).
2. Moving away from expecting Webpack to auto-shim node dependencies and instead requiring exactly what we need and depending on it in that way, e.g. users should be able to `npm i @react-pdf/renderer`, then run in Vite, Webpack 5, and Node without issue.

@carlobeltrame If you have time to work on a PR, we're using React PDF in both Vite and Node, so I can test server side (and create an example project) for you if that helps.

**Comment by Belkacem** [comment]
@thekevinbrown it doesn't seem to work with a fresh install (aka delete yarn lock file and yarn install)
tried on your published examples repository.
Thanks !

**Comment by carlobeltrame** [comment]
@Belkacem did you also insert the shim into the vite config?

**Comment by Belkacem** [comment]
> @Belkacem did you also insert the shim into the vite config?

Yeah, we did that, the [example](https://github.com/TechInSite/react-pdf-development-harness) provided by @thekevinbrown works perfectly until you do an clean install (delete yarn.lock and yarn install in the vite directory), I just tried it it spits `ReferenceError: Buffer is not defined`

**Comment by carlobeltrame** [comment]
Do you have a stack trace where exactly that happens? Maybe some of the recent new code added another usage of `Buffer` and @thekevinbrown and me are not yet using the newest version so we don't notice.
Also, when working with Vite and changing a lot in the dependencies, I sometimes have to stop the Vite dev server, delete the `node_modules/.vite` directory and start the dev server again. Otherwise Vite will sometimes keep on using old cached (pre-bundled) versions of some dependencies.

**Comment by Belkacem** [comment]
It is happening on a fresh clone, I cleaned the yarn cache before testing and it's happening consistently, here's a the stack trace : 
`Uncaught ReferenceError: Buffer is not defined
    at node_modules/@react-pdf/unicode-properties/lib/unicode-properties.es.js (unicode-properties.es.js:268:31)
    at __init (chunk-GVC7CBB2.js?v=33aefd43:15:56)
    at node_modules/@react-pdf/fontkit/lib/fontkit.browser.es.js (fontkit.browser.es.js:10:1)
    at __init (chunk-GVC7CBB2.js?v=33aefd43:15:56)
    at node_modules/@react-pdf/font/lib/index.browser.js (fontkit.browser.es.js:40110:29)
    at __require2 (chunk-GVC7CBB2.js?v=33aefd43:18:50)
    at asyncToGenerator.js:35:1`

**Comment by carlobeltrame** [comment]
Actually I just realized I have the same problem when using the latest version of react-pdf. It even was one of the reasons that motivated me to start working on #1891. It seems to have started after updating to react-pdf 2.1.2, but might have been sooner.

I'll see if I find a quick fix for the shim. But otherwise, I hope we can either merge #1891 or follow Devon Govett's idea there to return to the (soon isomorphic) foliojs packages instead of forking them for react-pdf.

**Comment by carlobeltrame** [comment]
Analyzing further, it must have come from this change: https://github.com/diegomura/react-pdf/pull/1839
Fix for the shim is here: https://github.com/exogee-technology/vite-plugin-shim-react-pdf/pull/8
Or for now you can use @react-pdf/renderer < 2.1.2 @Belkacem

**Comment by thekevinbrown** [comment]
This has now been released, sorry for the delay!

**Comment by Keshish** [comment]
Any idea Regarding,

```
> [vite]: Rollup failed to resolve import "file-loader!pdfjs-dist/legacy/build/pdf.worker" from "node_modules/react-pdf/dist/esm/entry.webpack.js".
> This is most likely unintended because it can break your application at runtime.
> If you do want to externalize this module explicitly add it to
> `build.rollupOptions.external`
> error during build:
> Error: [vite]: Rollup failed to resolve import "file-loader!pdfjs-dist/legacy/build/pdf.worker" from "node_modules/react-pdf/dist/esm/entry.webpack.js".
> This is most likely unintended because it can break your application at runtime.
> If you do want to externalize this module explicitly add it to
> `build.rollupOptions.external`
```

**Comment by carlobeltrame** [comment]
@Keshish no idea, never seen that error. Please open a new issue and provide a way to reproduce your problem.

**Comment by krishna-404** [comment]
The shim patch doesnt install with vite > 3.0, its dependency is vite 2. Any plans on working on it?

Have been trying to work with this for the last 2 days & issues go back all the way to pdfkit & the libraries used by pdfkit. React-pdf needs a complete overhaul to make it work with react18 & workbox5. but how does all that get resolved unless the upward dependencies are not sorted.

I would love to contribute to this, but lack in knowledge depth presently. 

Here is the error. Trying --legacy-peer-deps, doesn't help :|

```
npm ERR! Found: vite@3.1.8
npm ERR! node_modules/vite
npm ERR!   dev vite@"^3.1.8" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer vite@"^2.0.0" from vite-plugin-shim-react-pdf@1.0.7
npm ERR! node_modules/vite-plugin-shim-react-pdf
npm ERR!   dev vite-plugin-shim-react-pdf@"*" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /Users/krishna404/.npm/eresolve-report.txt for a full report.
```

**Comment by mmospanenko** [comment]
3.3.8 pnpm, monorepo, the same issue. With was ok (when packages were hoisted)
 
<img width="600" alt="image" src="https://github.com/diegomura/react-pdf/assets/2766211/4a240fcb-7764-4bf6-857a-56fe4caf3a84">
<img width="508" alt="image" src="https://github.com/diegomura/react-pdf/assets/2766211/3c0fe248-3fe4-44db-8da8-be92ccf30199">
